### ISSUE TITLE
Filesystem watch/notification status per OS/arch

### ISSUE THREAD

**Issue by calmh** [syncthing/syncthing#4503]
So historically we mostly build releases with CGO_ENABLED=0 to get a static binary with minimum dependencies. This doesn't work for the fs watcher, at least not on Mac. So, check what the requirements are and check off that we're doing the right thing. Table below represents current status as of changes and tests.

| OS | CGO Enabled | FS Watcher | Comments |
|--------|---|-----------|----------|
| macOS amd64 | Yes | ✅ | works on 10.8 |
| macOS 386 | Yes | ✅ | works on 10.8 |
| Linux amd64   | No  | ✅ | |
| Linux 386   | No | ✅ | |
| Linux exotics  | No | ❔ | should work |
| Android, built as linux  | No | ✅ |  |
| VariousBSD cross compiled  | No | ❔ | should work |
| Solaris cross compiled  | No | :x: | requires CGO, native build |
| Windows amd64 | No | ✅ | |
| Windows 386 | No | ❔ | should work? |

Enabling CGO on Mac means we add a dependency on the system libraries. I'm not sure if is the version of the OS we are building on or the minimum supported by Go. There were previous issues on this...

Paging @imsodin / @AudriusButkevicius to see if the current master builds work on your operating systems of choice...

**Comment by AudriusButkevicius** [comment]
Builds fine on Windows without CGO.

**Comment by calmh** [comment]
Just to be super clear, you mean that the version built without CGO on build.syncthing.net works with fs notifications? (I know it *builds* without CGO, that's what the build server does.)

**Comment by AudriusButkevicius** [comment]
sec lemme try

**Comment by AudriusButkevicius** [comment]
So it starts, no errors, yet I am not sure it works.

**Comment by calmh** [comment]
Enable fs notifications on some folder and see log output. It should say "Started fs notifications for ..." or something

**Comment by AudriusButkevicius** [comment]
It works, no errors.

**Comment by calmh** [comment]
Cool

**Comment by calmh** [comment]
Can you try the 386 version as well same?

**Comment by AudriusButkevicius** [comment]
sec

**Comment by calmh** [comment]
Linux is fine with cgo disabled on the archs I can test

**Comment by AudriusButkevicius** [comment]
So 386 doesn't work for me at all, getting some random

```
c:\test>syncthing.exe -home c:\test
[monitor] 23:43:59 INFO: Log output saved to file "c:\test\syncthing.log"
[monitor] 23:43:59 INFO: Starting syncthing
[start] 23:43:59 INFO: Generating ECDSA key and certificate for syncthing...
[GAJ66] 23:43:59 INFO: syncthing v0.14.40+27-gc080f67 "Dysprosium Dragonfly" (go1.9.2 windows-386) teamcity@build.syncthing.net 2017-11-11 19:18:17 UTC
[GAJ66] 23:43:59 INFO: My ID: GAJ66SF-QGR7E3A-Y3XQICO-FVP4BTR-OTXHAVK-G6LACGH-ERZPHCM-SCDIHAI
[GAJ66] 23:44:00 INFO: Single thread SHA256 performance is 128 MB/s using crypto/sha256 (77 MB/s using minio/sha256-simd).
[GAJ66] 23:44:00 INFO: Default folder created and/or linked to new config
[GAJ66] 23:44:00 INFO: Default config saved. Edit c:\test\config.xml to taste or use the GUI
[GAJ66] 23:44:01 INFO: Hashing performance with weak hash is 108.43 MB/s
[GAJ66] 23:44:01 INFO: Hashing performance without weak hash is 127.79 MB/s
[GAJ66] 23:44:01 INFO: Weak hash enabled, as it has an acceptable performance impact.
[GAJ66] 23:44:01 FATAL: Cannot open database: Access is denied. - Is another copy of Syncthing already running?
[monitor] 23:44:01 INFO: Syncthing exited: exit status 1
```

**Comment by calmh** [comment]
Awesome. Maybe that doesn't work on non-386 windowses or something

**Comment by Nutomic** [comment]
Is the file watcher also expected to work on Android? If so, I'll give it a try tomorrow.

**Comment by imsodin** [comment]
@Nutomic syncthing-android is built with GOOS=android, right? In that case there is no support for it in the notify library. Their inotify watcher uses golang.org/x/sys/unix to access the inotify functionality of the kernel. If this works on android, it should be as easy to add support for it as adding one build flag. I assume there is a catch, otherwise I'd expect they would have done that already.

**Comment by AudriusButkevicius** [comment]
I suspect linux watcher should work on android I suspect.

**Comment by imsodin** [comment]
I asked upstream: rjeczalik/notify#133

**Comment by calmh** [comment]
> I suspect linux watcher should work on android.

Your naivety is charming.

**Comment by imsodin** [comment]
Apparently GOOS android uses the Linux build path, so if it works it already does. I don't have the new syncthing android version yet, so i can't test yet

**Comment by Nutomic** [comment]
Just tried it on Android 6.0 and 7.1, neither is working. Do I need to add a compile flag or anything? I didn't see anything in the logs (checked STTRACE flags filesystem and scanner).

@imsodin The Android app has basic inotify integrated, so you should recompile and comment that out if you want to test it yourself. [comment out the whole function except the onApiChange() call](https://github.com/syncthing/syncthing-android/blob/0d964b290d071f6c0a9aa6a000e96530fc1a072a/app/src/main/java/com/nutomic/syncthingandroid/service/SyncthingService.java#L216)

**Comment by imsodin** [comment]
@Nutomic You don't need to set any compile flag, you just need to st FSWatcherEnabled to true in the folder config.
Thanks for the pointer about what to comment out, but I haven compiled and deployed any android app and am not to keen on that right now to be honest.

**Comment by Nutomic** [comment]
Oh wait I actually forgot to set the config option... Let me try again.

**Comment by Nutomic** [comment]
Okay events work! Even on Android 6.0, where the Java API for inotify is broken.

However, deletes seem to be processed very slowly. Is this expected? Takes about one minute until a delete is forwarded to Syncthing, with the default setting for fs watcher delay. Creates are processed within ~10 seconds.

**Comment by AudriusButkevicius** [comment]
Deletes probably come from normal rescan

**Comment by imsodin** [comment]
Yes that is to be expected. That is a compromise to make sure that creations are always processed before deletions, such that moves/local reuse of block data can happen. Events are delayed if a file gets modified repeatedly, up to 1min (or 6*FSWatcherDelay if it is configured to be shorter than 10s). Therefore deletes are always delayed for the same time. These timings are no science, so if anyone has a reason to change them, just say so.

**Comment by AudriusButkevicius** [comment]
And renames are detected as create + delete right? What happens if both land in the same directory?

**Comment by imsodin** [comment]
The fs watch does not detect renames. It just schedules scans on created files before it schedules scans on deleted files. Actually that only makes sure block data can be copied, but prevents direct renames. Probably it would be better to schedule them together and just order them correctly. Also we are getting off-topic :P
